lib/Sema/TypeCheckType.cpp

```diff
@@ -827,23 +827,27 @@ static Type diagnoseUnknownType(TypeChecker &tc, DeclContext *dc,
       NominalTypeDecl *nominal = nullptr;
       if ((nominalDC = dc->getInnermostTypeContext()) &&
           (nominal = nominalDC->getAsNominalTypeOrNominalTypeExtensionContext())) {
-        // Attempt to refer to 'Self' within a non-protocol nominal
-        // type. Fix this by replacing 'Self' with the nominal type name.
-
-        // Retrieve the nominal type and resolve it within this context.
+        // Attempt to refer to 'Self' within a non-protocol nominal type
         assert(!isa<ProtocolDecl>(nominal) && "Cannot be a protocol");
-        auto type = resolver->mapTypeIntoContext(
-          dc->getInnermostTypeContext()->getSelfInterfaceType());
-        if (type->hasError())
-          return type;
 
         // Produce a Fix-It replacing 'Self' with the nominal type name.
         auto name = getDeclNameFromContext(dc, nominal);
         tc.diagnose(comp->getIdLoc(), diag::self_in_nominal, name)
           .fixItReplace(comp->getIdLoc(), name);
-        comp->overwriteIdentifier(nominal->getName());
-        comp->setValue(nominal, nominalDC->getParent());
-        return type;
+
+        // If this is a requirement, replacing 'Self' with a valid
+        // type will result in unnecessary diagnostics. Simply return
```

Comment:
I'd like to see a little bit more, about why the diagnostics are unnecessary, e.g. "because it is likely to be an invalid type to use in such places".